src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -121,8 +123,22 @@ def osd_task(name, metadata, wait_for=2.0):
 @APIRouter('/osd', Scope.OSD)
 @APIDoc('OSD management API', 'OSD')
 class Osd(RESTController):
-    def list(self):
-        osds = self.get_osd_map()
+    def list(self, offset: int = 0, limit: int = 10,
+             search: str = '', sort: str = ''):
+        all_osds = self.get_osd_map()
```

Comment:
We should probably work with the Core team (CC @rzarzynski) to extend the underlying call, `mgr.get('osd_map')` (and others with potential scalability issues) with a common interface to allow random access to lists, and, perhaps, search/sorting capabilities:
```suggestion
        all_osds = self.get_osd_map(offset=int(offset), limit=int(limit), sort=sort, ... )
```
Long time ago, we identified that the process of converting internal Ceph arrays to Python objects was heavily impacting mgr scalability. However, Dashboard and other modules usually require a small subset of data than the one provided by the `get_osd_map` call, which ultimately means a waste of resources.

This should be made available to other ceph-mgr API methods like `mgr.get('osd_stats')`, `mgt.get('osd_map_tree')`, etc.

The approach could involve extending `OSDMap::dump(Formatter *f, CephContext *cct)` to support some iterator/slice semantics (start/offset, end/limit, sort, etc). Instead of altering the `dump()` itself, we might add the pagination to the `Formatter` (this would be consistent with the similar approach of C++ `ranges`/`views`, separating presentation from data holding).

---

**Alternatively, we could follow the opposite approach** (proposed by @tchaikov) of creating smart CPython classes, like the `OsdMap`:
https://github.com/ceph/ceph/blob/17dd1a270d3adeea850f0af4d6bc0ee207fab5dd/src/pybind/mgr/mgr_module.py#L201-L267

However, that's more challenging as it requires both expertise on CPython and C++ .

Reply:
yeah, this would actually improve dashboard performance by a lot..